dali/kernels/common/join/tensor_join_cpu.h

```diff
@@ -0,0 +1,211 @@
+                 -ndims + 1, ", ", ndims - 1, "] interval");
+    axis_ = axis >= 0 ? axis : ndims + axis;
+
+    {
+      const auto &ref = in_shapes[0];
+      for (int i = 1; i < n_input_tensors_; i++) {
+        DALI_ENFORCE(in_shapes[i].sample_dim() == ref.sample_dim(),
+                     "Every input shape must have the same number of dimensions.");
+        for (int j = 0; j < ref.size(); j++) {
+          if (!new_axis) {
+            DALI_ENFORCE(in_shapes[i][j] == ref.shape[j] || j == axis_, make_string(
+                    "Number of samples in every dimension (except the concatenated one) "
+                    "must be the same (CONCAT mode). 0-th shape at index ", j, " has dimension ",
+                    ref.shape[j], ", while ", i, "-th shape at index ", j,
+                    " has dimension ", in_shapes[i][j]));
+          } else {
+            DALI_ENFORCE(in_shapes[i][j] == ref.shape[j], make_string(
+                    "Number of samples in every dimension must be the same (STACK mode). "
+                    "0-th shape at index ", j, " has dimension ", ref.shape[j], ", while ", i,
+                    "-th shape at index ", j, " has dimension ", in_shapes[i][j]));
+          }
+        }
+      }
+    }
+
+    KernelRequirements kr;
+    output_shape_ = DetermineShape<new_axis>(in_shapes, axis);
+    kr.output_shapes.resize(1);
+    TensorListShape<> tmp({output_shape_});  // clang's destructor bug still haunting
+    kr.output_shapes[0] = tmp;
+    return kr;
+  }
+
+
+  KernelRequirements Setup(KernelContext &ctx, span<const InTensorCPU<T, dims>> in, int axis) {
+    std::vector<TensorShape<>> in_shapes(in.size());
+    for (int i = 0; i < in.size(); i++) {
+      in_shapes[i] = in[i].shape;
+    }
+    return Setup(ctx, make_span(in_shapes), axis);
+  }
+  ///@}
+
+  static constexpr int output_dims = (dims == DynamicDimensions ? DynamicDimensions :
+                                      (new_axis ? dims + 1 : dims));
+
+
+  /**
+   * @param out output tensor. Must be properly allocated
+   * @param in input tensors. The number of these tensors, as well as their shapes,
+   *           must match with what is provided in Setup call.
+   */
+  void Run(KernelContext &ctx, const OutTensorCPU<T, output_dims> &out,
+           span<const InTensorCPU<T, dims>> in) {
+    if (in.size() != n_input_tensors_) {
+      throw std::invalid_argument(make_string(
+              "Input must have the same number of tensors as was specified in call to Setup. "
+              "Expected: ", n_input_tensors_, "Actual: ", in.size()));
+    }
+    for (int i = 0; i < n_input_tensors_; i++) {
+      if (in[i].shape != orig_shapes_[i]) {
+        throw std::invalid_argument(make_string(
+                "Input must have the same shapes as was specified in call to Setup. Expected: ",
+                orig_shapes_[i], "Actual: ", in[i].shape));
+      }
+    }
```

Comment:
This bothers me. We're keeping the original shapes just to run this extremely defensive check.

Reply:
The one check that _would_ be useful is missing, though: check that the output actually has the size that we've calculated in Setup.
Rationale:
Not checking the *inputs* is acceptable since it's unlikely that they've changed between Setup and Run. The caller must have taken the shapes from somewhere and it's highly unlikely that they won't match.
Not checking the *output* is dangerous, because we've never seen the output tensor before and there's a chance that the caller simply forgot to resize it. This runs the risk of corrupting memory.